src/slurm_plugin/clustermgtd.py

```diff
@@ -407,19 +394,20 @@ def _handle_successfully_launched_nodes(self, healthy_nodes):
         Includes resetting partition bootstrap failure count for these nodes type and updating the
         bootstrap_failure_nodes_types.
         If a node has been failed successfully launched, the partition count of this node will be reset.
-        So the successfully launched nodes type will be remove from bootstrap_failure_nodes_types. If there's node types
-        failed in this partition later, it will keep count again.
+        So the successfully launched nodes type will be remove from bootstrap_failure_nodes_types.
+        If there's node types failed in this partition later, it will keep count again.
         """
-        online_nodes_types = self._get_online_nodes_types(healthy_nodes)
+        # Find nodes types which have been failed during bootstrap now become online.
```

Comment:
The `_handle_successfully_launched_nodes` could be optimized. Why even bothering doing anything if the failure count for the partition is already 0 (which I expect it to be most of the times).

Also what we are doing now is to reset the failure counter only if a node of a failed type is successfully launched. However we have a threshold on the entire partition regardless of the type. I'm  wondering at this point if it makes sense to simply reset the partition counter when at least one node is online in the partition.

I'd propose something like this (this will be even simpler if we don't want to discriminate by type):
1. rather than storing `_bootstrap_failure_nodes_types` and `_partitions_protected_failure_count_map` we only store `_partitions_protected_failure_count_map` but we make it store failures by compute resource type like the following (note that in case we don't want to discriminate between instance types we can have a single counter per partition):
```
{
   "partition1": { "compute_resource_1": 0, "compute_resource2": 2, ...},
   ...
}
```
2. Then we change the logic to something like the following pseudocode:
```
for partition, failures_per_compute_resource in _partitions_protected_failure_count_map.items():
    if sum(failures_per_compute_resource.values()) > 0:
        for compute_resource, failure_count in failures_per_compute_resource.items():
             if failure_count > 0:
                 if partitions[partition].has_online_node_by_type(compute_resource):  # note that has_online_node_by_type will be a method from SlurmParition
                     reset_partition_failure_count(partition)
                     break
```